pubsub/aws/snssqs/snssqs.go

```
@@ -526,55 +509,89 @@ func (s *snsSqs) createQueueAttributesWithDeadLetters(queueInfo, deadLettersQueu
 	return sqsSetQueueAttributesInput, nil
 }
 
+func (s *snsSqs) restrictQueuePublishPolicyToOnlySNS(sqsQueueInfo *sqsQueueInfo, snsARN string) error {
+	// only permit SNS to send messages to SQS using the created subscription.
+	if _, err := s.sqsClient.SetQueueAttributes(&(sqs.SetQueueAttributesInput{
+		Attributes: map[string]*string{
+			"Policy": aws.String(fmt.Sprintf(`{
+				"Version": "2012-10-17",
+				"Statement": [{
+					"Effect":"Allow",
+					"Principal":{"Service": "sns.amazonaws.com"},
+					"Action":"sqs:SendMessage",
+					"Resource":"%s",
+					"Condition": {
+						"ArnEquals":{
+						  "aws:SourceArn":"%s"
+						}
+					  }	
+				}]
+			}`, sqsQueueInfo.arn, snsARN)),
+		},
+		QueueUrl: &sqsQueueInfo.url,
+	})); err != nil {
+		return fmt.Errorf("error setting queue subscription policy: %w", err)
+	}
+
+	return nil
+}
+
 func (s *snsSqs) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handler) error {
-	// subscribers declare a topic ARN
-	// and declare a SQS queue to use
-	// these should be idempotent
-	// queues should not be created if they exist
+	// subscribers declare a topic ARN and declare a SQS queue to use
+	// these should be idempotent - queues should not be created if they exist.
 	topicArn, err := s.getOrCreateTopic(req.Topic)
 	if err != nil {
-		s.logger.Errorf("error getting topic ARN for %s: %v", req.Topic, err)
+		s.logger.Errorf("error getting topic ARN for %s: %w", req.Topic, err)
```

Wrapped error is not returned